Add standard Ruby collection methods to Result and Row

wpfleger96 opened the pull request:
This PR adds standard Ruby collection methods for query results: `empty?`, `dig`, `key?`, and `fetch`. A recent BigQuery to Snowflake migration revealed these methods were missing; `Result` includes `Enumerable` but lacks `empty?`, while `Row` has `[]` access but no safe nested access or column existence checks.

Implementation:

- Add `Result#empty?` to check if result set has no rows
- Add `Row#dig` for safe nested value access (useful for VARIANT/JSON columns)
- Add `Row#key?`/`has_key?` to check column existence
- Add `Row#fetch` with default values and block support
- All methods maintain case-insensitive column access

wpfleger96 committed:
Add standard Ruby collection methods to Result and Row

- Add Result#empty? for checking empty result sets
- Add Row#dig for safe nested value access (useful for VARIANT/JSON columns)
- Add Row#to_h for explicit hash conversion
- Add Row#key?/has_key? for column existence checks
- Add Row#fetch for access with default values or error handling
- Add unit tests (28 test cases, no Snowflake required)

Addresses missing methods discovered during BigQuery to Snowflake migration.

reidnimz commented:
Thanks @wpfleger96 - I'll try to steal some time and take a look through this in the next week or so, much appreciated!

wpfleger96 commented:
@reidnimz 😅 I think the tests are failing bc GH Actions is blocking secrets since I opened this PR from a fork

reidnimz commented on lib/ruby_snowflake/row.rb:
```diff
@@ -77,6 +77,30 @@ def values
 
+    def to_h
+      @column_to_index.each_with_object({}) { |(name, _), hash| hash[name] = self[name] }
+    end
```
I think we should probably keep the Enumerable implementation of to_h here. 

If not, we'd probably want it to be slightly modified:
```
def to_h
  @column_to_index.each_with_object({}) { |(name, index), hash| hash[name] = self[index] }
end
```

But, I think it's probably better to just rely on the Enumerable implementation

wpfleger96 replied:
Great call! removed in [c9ecd2e71543fbe5b8f04bca08fb2cb5bbdd2da2](https://github.com/rinsed-org/rb-snowflake-client/pull/159/changes/c9ecd2e71543fbe5b8f04bca08fb2cb5bbdd2da2?new_files_changed=true)

reidnimz commented on CHANGELOG.md:
```diff
@@ -6,6 +6,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
+### Added
+- `Result#empty?` - Check if result set has no rows
+- `Row#dig` - Safe nested value access for VARIANT/JSON columns
+- `Row#to_h` - Explicit hash conversion (was implicit via Enumerable)
```
then we'd want to remove just this line if we rely on the Enumerable implementation

wpfleger96 replied:
Removed in [c9ecd2e71543fbe5b8f04bca08fb2cb5bbdd2da2](https://github.com/rinsed-org/rb-snowflake-client/pull/159/changes/c9ecd2e71543fbe5b8f04bca08fb2cb5bbdd2da2?new_files_changed=true)

reidnimz commented on spec/ruby_snowflake/row_spec.rb:
```diff
@@ -0,0 +1,53 @@
+      expect(row.fetch("missing", "default")).to eq("default")
+      expect(row.fetch(:nonexistent, 0)).to eq(0)
+    end
+
```
Should probably have the happy path in here too so we don't accidentally break it some day

wpfleger96 replied:
Added happy path test case in [c9ecd2e71543fbe5b8f04bca08fb2cb5bbdd2da2](https://github.com/rinsed-org/rb-snowflake-client/pull/159/changes/c9ecd2e71543fbe5b8f04bca08fb2cb5bbdd2da2?new_files_changed=true)

reidnimz commented on spec/ruby_snowflake/row_spec.rb:
```diff
@@ -0,0 +1,53 @@
+      expect { row.fetch("missing") }.to raise_error(KeyError, /key not found/)
+      expect { row.fetch(:nonexistent) }.to raise_error(KeyError, /key not found/)
+    end
+  end
```
Just need some simple test cases for `key?` as well since it's part of the API

wpfleger96 replied:
Added a few simple tests for `Row#key?` in [c9ecd2e71543fbe5b8f04bca08fb2cb5bbdd2da2](https://github.com/rinsed-org/rb-snowflake-client/pull/159/changes/c9ecd2e71543fbe5b8f04bca08fb2cb5bbdd2da2?new_files_changed=true)

reidnimz reviewed (changes requested):
Good addition, thanks @wpfleger96! Just a couple of minor things.

You're right, I forgot with the secrets, it's not going to work on a branch. But, I'll run it locally quick before confirming and then I'll try and get a release out today or tomorrow.

wpfleger96 committed:
addressing PR feedback

1. removed custom unnecessary to_h method implementation (and update CHANGELOG accordingly)
2. Add happy path test for Row#fetch
3. Added test cases for Row#key? since it's part of the API
